File: objective-c/ArcGISSample/Legend/LegendViewController.m
//
//  LegendViewController.m
//  ArcGISSample
//
//  Created by esrij on 2015/07/27.
//  Copyright (c) 2015年 esrij. All rights reserved.
//

#import "LegendViewController.h"


@implementation LegendViewController


- (void)viewDidLoad
{
    [super viewDidLoad];
    
    self.agsMapView = [[AGSMapView alloc] initWithFrame:self.view.bounds];
    [self.agsMapView enableWrapAround];
    [self.view addSubview:self.agsMapView];
    
    //タイルマップサービスレイヤーの追加
    NSURL *url = [NSURL URLWithString:@"http://server.arcgisonline.com/ArcGIS/rest/services/World_Street_Map/MapServer"];
    AGSTiledMapServiceLayer *agsTiledMapServiceLayer = [[AGSTiledMapServiceLayer alloc] initWithURL:url];
    [self.agsMapView addMapLayer:agsTiledMapServiceLayer withName:@"Tiled layer"];
    
    //凡例用のダイナミックマップサービスレイヤーの追加
	NSURL *dynamicLayerUrl = [NSURL URLWithString:@"http://sampleserver1.arcgisonline.com/ArcGIS/rest/services/Specialty/ESRI_StateCityHighway_USA/MapServer"];
    AGSDynamicMapServiceLayer *agsDynamicMapServiceLayer = [[AGSDynamicMapServiceLayer alloc] initWithURL:dynamicLayerUrl];
    [self.agsMapView addMapLayer:agsDynamicMapServiceLayer withName:@"Legend Layer"];
    
    agsDynamicMapServiceLayer.delegate = self;

    
}


- (void) layerDidLoad:(AGSLayer*) layer{

    if ([layer.name isEqualToString:@"Legend Layer"]) {
        
        [self.agsMapView zoomToEnvelope:layer.initialEnvelope animated:YES];
        
        //ダイナミックマップサービスレイヤーの表示設定（ID:2のレイヤを表示）
        AGSDynamicMapServiceLayer *dlayer = (AGSDynamicMapServiceLayer *)layer;
        dlayer.visibleLayers = @[@2];
        
        //ダイナミックマップサービスレイヤーの凡例情報の取得
        AGSMapServiceInfo *agsMapServiceInfo = dlayer.mapServiceInfo;
        agsMapServiceInfo.delegate = self;
        [agsMapServiceInfo retrieveLegendInfo];
        
        NSLog(@"%@", agsMapServiceInfo.serviceDescription);
        
    }
    
}


- (void)mapServiceInfo:(AGSMapServiceInfo *)mapServiceInfo operationDidRetrieveLegendInfo:(NSOperation *)op {
    
    
    UIScrollView *legendView = [[UIScrollView alloc] initWithFrame:CGRectMake(10, 100, 150, 200)];
    legendView.backgroundColor = [UIColor whiteColor];
    legendView.alpha = 0.8;
    [[legendView layer] setCornerRadius:10.0];
    

    double y = 10;

    //ID:2の凡例情報を取得
    AGSMapServiceLayerInfo *layerInfo = [mapServiceInfo.layerInfos objectAtIndex:2];
    
    for(int i=0; i<layerInfo.legendImages.count; i++){
        
        //各凡例の画像とラベルを取得
        UIImage *legendImage = [layerInfo.legendImages objectAtIndex:i];
        UIImageView *imageView = [[UIImageView alloc] initWithImage:legendImage];
        imageView.frame = CGRectMake(10, y, legendImage.size.width, legendImage.size.height);
        [legendView addSubview:imageView];
            
        UILabel *legendlabel = [[UILabel alloc] initWithFrame:CGRectMake(legendImage.size.width + 15, y, 100, legendImage.size.height)];
        legendlabel.font = [UIFont boldSystemFontOfSize:10];
        legendlabel.textColor = [UIColor blackColor];
        legendlabel.text = [layerInfo.legendLabels objectAtIndex:i];
        [legendView addSubview:legendlabel];

        y += legendImage.size.height;
            
    }
    
    legendView.contentSize = CGSizeMake(150, y);
    [self.view addSubview:legendView];
    
}


- (void)mapServiceInfo:(AGSMapServiceInfo *)mapServiceInfo operation:(NSOperation *)op didFailToRetrieveLegendInfoWithError:(NSError *)error {
    
    NSLog(@"Error: %@", error);
    
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}



@end

File: swift/ArcGISSample/LegendViewController.swift
//
//  LegendViewController.swift
//  ArcGISSample
//
//  Created by esrij on 2015/07/27.
//  Copyright (c) 2015年 esrij. All rights reserved.
//



import UIKit
import ArcGIS


class LegendViewController: UIViewController, AGSLayerDelegate, AGSMapServiceInfoDelegate {
    
    var agsMapView: AGSMapView!
    
    
    override func viewDidLoad() {
        
        super.viewDidLoad()
        
        agsMapView = AGSMapView(frame: view.bounds)
        agsMapView.enableWrapAround()
        view.addSubview(agsMapView)
        
        //タイルマップサービスレイヤーの追加
        let url = URL(string: "https://server.arcgisonline.com/ArcGIS/rest/services/World_Street_Map/MapServer")
        let tiledLyr = AGSTiledMapServiceLayer(url:url)
        agsMapView.addMapLayer(tiledLyr, withName:"Tiled Layer")
        
        //凡例用のダイナミックマップサービスレイヤーの追加
        let dynamicLayerUrl = URL(string: "https://sampleserver1.arcgisonline.com/ArcGIS/rest/services/Specialty/ESRI_StateCityHighway_USA/MapServer")
        let agsDynamicMapServiceLayer = AGSDynamicMapServiceLayer(url:dynamicLayerUrl)
        agsMapView.addMapLayer(agsDynamicMapServiceLayer, withName:"Legend Layer")

        agsDynamicMapServiceLayer?.delegate = self

    }
    
    func layerDidLoad(_ layer: AGSLayer!) {
        
        if (layer.name == "Legend Layer") {
            
            agsMapView.zoom(to:layer.initialEnvelope, animated: false)
            
            //ダイナミックマップサービスレイヤーの表示設定（ID:2のレイヤを表示）
            let dlayer = layer as! AGSDynamicMapServiceLayer
            dlayer.visibleLayers = [2]
            
            //ダイナミックマップサービスレイヤーの凡例情報の取得
            let agsMapServiceInfo = dlayer.mapServiceInfo
            agsMapServiceInfo?.delegate = self
            _ = agsMapServiceInfo?.retrieveLegendInfo()
            
            print("\(agsMapServiceInfo?.serviceDescription)")
            
        }
    }
    
    
    func mapServiceInfo(_ mapServiceInfo: AGSMapServiceInfo!, operation op: Operation!, didFailToRetrieveLegendInfoWithError error: Error!) {
        
        print("\(error.localizedDescription)")
    
    }
    
    
    func mapServiceInfo(_ mapServiceInfo: AGSMapServiceInfo!, operationDidRetrieveLegendInfo op: Operation!) {
        
        let legendView = UIScrollView(frame: CGRect(x: 10, y: 100, width: 150, height: 200))
        legendView.backgroundColor = UIColor.white
        legendView.alpha = 0.8
        legendView.layer.cornerRadius = 10.0
        
        var y:CGFloat = 10;
        
        //ID:2の凡例情報を取得
        let layerInfo = mapServiceInfo.layerInfos[2] as! AGSLayerInfo


        for i in 0 ..< layerInfo.legendImages.count  {
            //各凡例の画像とラベルを取得
            let legendImage = layerInfo.legendImages[i] as! UIImage
            let imageView = UIImageView(image: legendImage)
            let frame:CGRect = CGRect(x: 10, y: y, width:legendImage.size.width, height:legendImage.size.width)
            imageView.frame = frame
            legendView.addSubview(imageView)
            
            let legendlabel = UILabel(frame: CGRect(x: legendImage.size.width + 15, y: y, width: 100, height: legendImage.size.height))
            legendlabel.font = UIFont.boldSystemFont(ofSize: 10)
            legendlabel.textColor = UIColor.black
            legendlabel.text = layerInfo.legendLabels[i] as? String
            legendView.addSubview(legendlabel)
            
            y = y + legendImage.size.height
            
        }
        
        legendView.contentSize = CGSize(width: 150, height: y)
        view.addSubview(legendView)
        
    }
    
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
}
